### PR TITLE
[AscendNPU-IR]  Add PyTorch operator deployment for TileLang Ascend

### PR DESCRIPTION
**Pull request by yyyy999**
## Summary

Register TileLang Ascend operators to PyTorch with offline installation support.

## Changes

- **Operator Framework**: Add base operator class (`BaseOp`) with abstract methods for name, signature, kernel loading, implementation, and Python API
- **Operator Implementations**: 
  - `FlashAttentionOp`: Fixed shape (seq_len=512, dim=128) with online softmax algorithm
  - `GemmOp`: Dynamic shape GEMM supporting arbitrary M, N, K dimensions
- **Kernel Loader**: Standalone `NPUKernelLoader` that loads precompiled kernels without tilelang source code dependency
- **Operator Registry**: PyTorch operator registration with `torch.library.Library`
- **Precompile Script**: Compile and save kernels for offline distribution

## Usage

Three ways to call the operators:

```python
# Method 1: Package API
import tl_ascend_ops
output = tl_ascend_ops.flash_attention(q, k, v)
c = tl_ascend_ops.gemm(a, b)

# Method 2: Via torch_npu (requires tl_ascend_ops import first)
import tl_ascend_ops
import torch_npu
output = torch_npu.flash_attention(q, k, v)
c = torch_npu.gemm(a, b)

# Method 3: Via torch.ops
import tl_ascend_ops
output = torch.ops.tl_ascend_ops.flash_attention(q, k, v)
torch.ops.tl_ascend_ops.gemm(a, b, c)
```

## Dependencies

- torch >= 2.6.0
- torch_npu

### CODE REVIEW

File: tilelang_ascend_ops/README.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+- ✅ **独立运行时** - 不依赖 tilelang 源码
+- ✅ **PyTorch 集成** - 支持 `torch.ops.tl_ascend_ops.xxx` 调用
+- ✅ **torch_npu 注入** - 支持 `torch_npu.xxx` 调用
+
+## 安装
+
+### 环境要求
+
+- Python >= 3.8
+- PyTorch >= 2.0.0
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/README.md]: We need torch_npu >= 2.6.0

File: tilelang_ascend_ops/setup.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+"""
+TileLang Ascend Operators - Installation Configuration
+
+Pure Python package, dependencies:
+- torch >= 2.0.0
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/setup.py]: 2.6.0, 2.7.1 we dont test 2.0.0 yet

File: tilelang_ascend_ops/setup.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        "tl_ascend_ops.kernels",
+    ],
+    package_dir={"tl_ascend_ops": "src"},
+    package_data={
+        "tl_ascend_ops": ["kernels/**/*", "*.so", "*.pkl"],
+    },
+    include_package_data=True,
+    python_requires=">=3.8",
+    install_requires=[
+        "torch>=2.0.0",
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/setup.py]: change this 2.6.0

File: tilelang_ascend_ops/setup.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+VERSION = "0.1.0"
+
+readme_path = Path(__file__).parent / "README.md"
+long_description = readme_path.read_text(encoding="utf-8") if readme_path.exists() else ""
+
+setup(
+    name=PACKAGE_NAME,
+    version=VERSION,
+    author="TileLang Ascend Team",
+    author_email="tilelang@example.com",
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/setup.py]: do not use this

File: tilelang_ascend_ops/src/ops/flash_attention.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            raise ValueError("V must be an NPU tensor")
+        
+        seq_len, dim = Q.shape
+        if K.shape != (seq_len, dim) or V.shape != (seq_len, dim):
+            raise ValueError(f"Shape mismatch: Q={Q.shape}, K={K.shape}, V={V.shape}")
+        
+        if scale is None:
+            scale = (1.0 / dim) ** 0.5
+        
+        return torch.ops.tl_ascend_ops.flash_attention(Q, K, V, scale)
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/src/ops/flash_attention.py]: In flash attention example, compile/kernels/flash_attention.py, there are only three inputs.
The examples you provided here don't seem to match.

File: tilelang_ascend_ops/src/__init__.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
+"""
+TileLang Ascend Operators Package
+
+Provides PyTorch operator interfaces with offline installation support.
+
+Dependencies:
+- torch >= 2.0.0
```
Comment (CeleNewYear) [inline tilelang_ascend_ops/src/__init__.py]: change to torch >= 2.6.0

File: examples/torch_tl_ops/src/loader.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+def replace_by_longest_key(calculate_str: str, replace_dict: dict) -> str:
+    """Replace variable names in string"""
+    sorted_keys = sorted(replace_dict.keys(), key=lambda x: (-len(x), x))
+    result = calculate_str
+    for key in sorted_keys:
+        result = result.replace(key, str(replace_dict[key]))
+    return result
+
+
+class NPUKernelLoader:
```
Comment (CeleNewYear) [inline examples/torch_tl_ops/src/loader.py:43]: If the workspace derivation is not handled, please mark this as TODO and switch to cache-based mode once the cache is stable.

**Review by CeleNewYear — CHANGES_REQUESTED** [review]

**Review by CeleNewYear — APPROVED** [review]
LGTM
